pkg/image/api/types.go

```diff
@@ -60,6 +77,41 @@ type ImageLayer struct {
 	Size int64
 }
 
+// SignatureValidationResult holds a result of signature validation.
+type SignatureValidationResult string
+
+// ImageSignature holds a signature of an image.
+type ImageSignature struct {
+	// Describes a type of stored blob. Provided by caller.
+	Type string
+	// An opaque binary string which is an image's signature.
+	Content []byte
+	// A time of last server-side validation.
+	ValidatedOn unversioned.Time
+	// Result of validation performed at ValidatedOn.
+	ValidationResult SignatureValidationResult
+
+	// At most one of the following pointers can point to details. If the signature type isn't recognized,
+	// they will all be nil.
+
+	// Details for signature of type "atomic".
+	AtomicSignatureDetails *AtomicSignatureDetails
```

Comment:
So using Jessica and the console as a stand-in, here's what I would expect:

Signatures are submitted by the client and *interpreted* by one or more
trusted platform interpreters for signatures.  Today there is exactly one -
atomic.
Admins may selective enable / disable signature interpreters.  By default
we would enable atomic.
The interpretation of a signature is a bit of specific code that transforms
a completely arbitrary signature into useful data for clients and API
consumers.
By virtue of being platform registered, it ensures a higher level of
trust.  I.e., random users cannot set that data.
This data is the equivalent of the "Signed by Red Hat Inc" that appears in
an x509 cert.  Clients should easily be able to present that info to users.

The fields describe above were very reasonable:

1. Signer / signatory - the entity the user should be expected to trust
2. The item being signed - the image digest aka ImageName
3. A source URL / value that loosely identifies the signing value -
ImageLocation or some similar value (not DockerImageReference if at all
possible, esp. if it does not follow DIR rules)
4. The expiry, if specified
5. The signing date, if specified.

The goal is to communicate to a platform consumer.


On Mon, May 23, 2016 at 4:36 PM, Clayton Coleman <ccoleman@redhat.com>
wrote:

> The point is that the value you are exposing the API has to be useful for
> communicating to users.  Only values that have use for end users should be
> exported, and that would be the fields that are common enough that a
> clients can convey meaningful info.   Creating one struct type for all
> possible types does not accomplish that.  Defining a mapping between the
> signature data and the output struct does.
>
> On Thu, May 19, 2016 at 10:36 AM, Miloslav Trmač <notifications@github.com
> > wrote:
>
>> In pkg/image/api/types.go
>> <https://github.com/openshift/origin/pull/8371#discussion_r63889611>:
>>
>> > +// ImageSignature holds a signature of an image.
>> > +type ImageSignature struct {
>> > +	// Describes a type of stored blob. Provided by caller.
>> > +	Type string
>> > +	// An opaque binary string which is an image's signature.
>> > +	Content []byte
>> > +	// A time of last server-side validation.
>> > +	ValidatedOn unversioned.Time
>> > +	// Result of validation performed at ValidatedOn.
>> > +	ValidationResult SignatureValidationResult
>> > +
>> > +	// At most one of the following pointers can point to details. If the signature type isn't recognized,
>> > +	// they will all be nil.
>> > +
>> > +	// Details for signature of type "atomic".
>> > +	AtomicSignatureDetails *AtomicSignatureDetails
>>
>> These are generic details core to the platform (image name).
>>
>> No, these are specific to the signature format and the signature
>> semantics.
>>
>> The signature format explicitly calls this a “Docker reference” (for
>> which I was told “image name” is the equivalent OpenShift term). If a
>> future signature format will instead sign claims that an image identity is
>> an URL, the signature would contain
>> "identity":{"some-kind-of-url":"$value"} and the parsed contents of the
>> signature could contain this URL, but definitely not in the “Atomic
>> signature image name” field.
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly or view it on GitHub
>> <https://github.com/openshift/origin/pull/8371/files/a49f7a114201f3d8868dba2c837f84b85045e79c#r63889611>
>>
>
>